File: src/main/java/org/simbrain/util/math/ProbDistributions/TwoValued.java
// package org.simbrain.util.math.ProbDistributions;
//
// import org.simbrain.util.UserParameter;
// import org.simbrain.util.math.ProbabilityDistribution;
//
// /**
//  * Returns one of two values.
//  * TODO: Do this in terms of proper theory
//  * TODO: Interact with polarity
//  */
// public class TwoValued extends ProbabilityDistribution {
//
//     @UserParameter(
//             label = "Upper value",
//             order = 1)
//     private double upper = -1;
//
//     @UserParameter(
//             label = "Lower value",
//             order = 2)
//     private double lower = 1;
//
//     @UserParameter(
//             label = "Threshold",
//             description = "If >= this threshold choose upper, else lower",
//             order = 3)
//     private double p = .5;
//
//     /**
//      * Public constructor for reflection-based creation. You are encourage to use
//      * the builder pattern provided for ProbabilityDistributions.
//      */
//     public TwoValued() {
//     }
//
//     public double nextDouble() {
//         return Math.random() > p ? upper : lower;
//     }
//
//     public int nextInt() {
//         return (int) nextDouble();
//     }
//
//
//     public String getName() {
//         return "Two valued";
//     }
//
//     public String toString() { return "Two valued"; }
//
//     @Override
//     public TwoValued deepCopy() {
//         TwoValued cpy = new TwoValued();
//         cpy.upper = this.upper;
//         cpy.lower = this.lower;
//         cpy.p = this.p;
//         return cpy;
//     }
//
//     @Override
//     public void setClipping(boolean clipping) {
//     }
//
//     @Override
//     public void setCeiling(double ceiling) {
//         this.upper = ceiling;
//     }
//
//     @Override
//     public void setFloor(double floor) {
//         this.lower = floor;
//     }
//
// }

File: src/main/kotlin/org/simbrain/util/stats/distributions/TwoValued.kt
package org.simbrain.util.stats.distributions;

import org.simbrain.util.UserParameter
import org.simbrain.util.stats.ProbabilityDistribution

/**
 * Returns one of two values based on a probability. A Bernoulli distribution where the sample space can be any two
 * numbers.
 */
class TwoValued(
    @UserParameter(label = "Lower value", useSetter = true, order = 1)
    var lowerValue: Double = -1.0,

    @UserParameter(label = "Upper value", useSetter = true, order = 2)
    var upperValue: Double = 1.0,

    @UserParameter(
        label = "Probability",
        description = "Probabiliy of selecting the upper value",
        order = 3)
    var p: Double = .5
) : ProbabilityDistribution() {

    override fun sampleDouble(): Double {
        return if (randomGenerator.nextDouble() > p) lowerValue else upperValue
    }

    override fun sampleInt(): Int {
        return sampleDouble().toInt()
    }

    override fun sampleDouble(n: Int): DoubleArray {
        return DoubleArray(n) {sampleDouble()}
    }

    override fun sampleInt(n: Int): IntArray {
        return IntArray(n) {sampleInt()}
    }

    val mean get() = p * upperValue + (1-p) * lowerValue

    val variance get() = .5 * (p * p + ((1-p) * (1-p))) * ((upperValue - lowerValue) * (upperValue - lowerValue))

    override fun deepCopy(): TwoValued {
        val copy = TwoValued()
        copy.randomSeed = randomSeed
        copy.lowerValue = lowerValue
        copy.upperValue = upperValue
        copy.p = p
        return copy
    }

    override fun getName(): String {
        return "Two valued"
    }

}